sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala

```diff
@@ -191,11 +194,15 @@ private[sql] class SharedState(
 }
 
 object SharedState extends Logging {
-  try {
-    URL.setURLStreamHandlerFactory(new FsUrlStreamHandlerFactory())
-  } catch {
-    case e: Error =>
-      logWarning("URL.setURLStreamHandlerFactory failed to set FsUrlStreamHandlerFactory")
+  private def setFsUrlStreamHandlerFactory(conf: SparkConf): Unit = {
+    if (conf.get(DEFAULT_URL_STREAM_HANDLER_FACTORY_ENABLED)) {
+      try {
+        URL.setURLStreamHandlerFactory(new FsUrlStreamHandlerFactory())
```

Comment:
This will set it each time SharedState is instantiated; do you still want a flag to remember whether it has been set? I don't know if it matters as well that multiple SharedState instances will share whatever this sets.

Reply:
The following is `setURLStreamHandlerFactory` of `URL.java`. Java already tracks it with `factory` variable. I don't think our provide any benefit. So, invoking `setURLStreamHandlerFactory` multiple times is okay.
```
    public static void setURLStreamHandlerFactory(URLStreamHandlerFactory fac) {
        synchronized (streamHandlerLock) {
            if (factory != null) {
                throw new Error("factory already defined");
            }
            SecurityManager security = System.getSecurityManager();
            if (security != null) {
                security.checkSetFactory();
            }
            handlers.clear();
            factory = fac;
        }
    }
```

Reply:
OK. It looks like it causes an error if called twice, so don't we need to guard against this? previously it would only be called once for sure as it was in an object's constructor.

Reply:
It's okay since we have `try .. catch`. And, `SharedState` creation is not performance critical code path.

Reply:
Oh, you mean that we are going to show the multiple warnings, right?

Reply:
It's possible. Actually, I guess we had better lower the log level from `logWarning` to `logInfo`.

Reply:
That, and simply relying on it being Error, so you have to handle it for control flow, etc.

Reply:
Sorry, I didn't understand what you mean. Do you want me to add the initial approach? I'm okay to add back if you want.
> That, and simply relying on it being Error, so you have to handle it for control flow, etc.

Reply:
I think that's why the initial PR hold the reference and check if it's set to prevent setting multiple time.

Reply:
Yep. Got it. No problem. `SharedState` is not created frequently, it's better to preserve the exiting behavior. I'll update once again.

Reply:
I mean, there's no reason to hold those references. But you may need an 'initialized' boolean, or check whether it has already been set.